README.md
# FurtherLand
A Blog Platform, written in Python, using Tornado and MongoDB, with fully asynchronous support.

Important
---------
FurtherLand is in development, DO NOT USE it in production environment unless you know what you are doing.

Requirements
------------
 - Python>=3.3(FurtherLand will not work under Python2)
 - Tornado>=4.0
 - Motor>=0.4
 - Feedgen>=0.3.1
 - Pykismet3>=0.1.1
 - bcrypt>=2.0.0
 - markdown>=2.6.2
 - py-gfm>=0.1.1
 - pyotp>=2.0.0

All listed Requirements(Except Python3.3) are able to be installed by pip.

License
-------
   Copyright 2015 Futur Solo

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.


CHANGELOG.md
# Changelog

v20150912 dev
-------------
 - Bugs Fix
 - Remove bundled pyotp library, use pyotp from PyPi(version>=2.0.0).
 - Two Factor Authentication Login Experience Optimization
 - Management Office Draft and Not Permitted Reply Statistics

v20150731 dev
-------------
 - Switch Markdown Rendering Library from Misaka to markdown with py-gfm extension

v20150725 dev
-------------
 - Bugs Fix
 - New FurtherLand Used Time Header

v20150724 dev
-------------
 - Bugs Fix

v20150723 dev
-------------
 - Bugs Fix

v20150722 dev
-------------
 - Removed useless console logs.

v20150721 dev
-------------
 - Bugs Fix
 - Polymer Framework Removed
 - New Management Office without Polymer
 - Remove GitHub OAuth Configuration from Library Sample
 - Summer Night Firefox and Safari browser Compatibility

v20150719 dev
-------------
 - New Theme Summer Night Released.

v20150716 dev
-------------
 - All Internal Areas has moved to New Terminal Service(A new, united API handler).

v20150715 dev
-------------
 - Bugs Fix
 - Management Office Security Enhancement
 - Mako Template has been removed since it is not more flexible than default tornado template but causes more problems.

v20150713 dev
-------------
 - Bugs Fix
 - Reply Area GitHub OAuth System removed, because it is not a graceful enough way to stop spammers.

v20150712 dev
-------------
 - Bugs Fix
 - New Checkin Office User Interface
 - Changed Password Encryption Algorithm to Blowfish Encrypt Algorithm

v20150711 dev
-------------
 - Bugs Fix
 - New Main Office with full pjax technology.
 - Checkin Progress Optimization
 - Rising Progress Optimization
 - Database Action Counts Removed

v20150707 dev
-------------
 - Removed Config Preload and FurtherLand SelfKill function because these are not stable enough.

v20150704 dev
-------------
 - Bugs Fix
 - Virtualenv has been removed from Requirements List, please use venv Python Standard Library as instead.
 - Formally Start to use Polymer Framework

v20150629 dev
-------------
 - More Graceful FurtherLand Rising Method
 - Polymer Framework Added

v20150531 dev
-------------
 - Stop Using Static File Content Delivery Network to Prevent Connection Problems

v20150529 dev
-------------
 - Bugs Fix
 - Stop Using Asyncio Event Loop Because it is slower than Tornado IO Loop
 - Reply Area GitHub OAuth System Works
 - Atom RSS Feed System Works

v20150502 dev
-------------
 - Bugs Fix
 - Stop Support for pypy3(Because it is slower than normal Python3 on Tornado) and update minimal Python version to 3.3
 - Master Information Preload Function
 - Public URL function Removed
 - Writings and Pages Erase Function
 - Fixed a serious Spelling Mistake

v20150501 dev
-------------
 - Bugs Fix
 - Performance Optimization
 - Config Preload Function
 - Template Preload Function
 - Footer Used Time and DB Actions Function

v20150430 dev
-------------
 - Bugs Fix
 - Reply Edit Function Works

v20150429 dev
-------------
 - Reply Management Pages has more details about the Replies
 - Disabled Mako Module Folder to Prevent 500 Errors in C10K Test

v20150424 dev
-------------
 - Anti Base64 and Javascript Protocol XSS Function
 - Use Asyncio Event Loop if Possible

v20150422 dev
-------------
 - Safe Land Option(HSTS, Secure Attributed Cookies, etc.)
 - Starts to Use Httponly Cookies

v20150420 dev
-------------
 - Bugs Fix
 - Moved Template Files to Factory Folder
 - NotFoundHandler Renamed to LostAndFoundPlace
 - SpiritPlace Removed
 - robots.txt Added

v20150419 dev
-------------
 - Bugs Fix
 - Management Office Working Page Preview Supports Code Highlights
 - Replies Supports Code Highlights
 - Management Office Replies Supports Code Highlights
 - Site Favicon Supports
 - Switch Content Delivery Network from Microsoft to CloudFlare
 - 404 Page Message Updates

v20150418 dev
-------------
 - Bugs Fix
 - Supports for Analytical Code(Such as Google Analytics)
 - Management Office Reply Details
 - Custom Error Pages

v20150417 dev
-------------
 - Bugs Fix
 - Supports for Public Library(Database Without Authentication)
 - Melody Options Updated

v20150414 dev
-------------
 - Bugs Fix

v20150413 dev
-------------
 - Bugs Fix
 - Working Management Function Works
 - Reply Management Function Works
 - Pages Funciton Works
 - Configuration Modify Function Works
 - Code Highlights Works

v20150412 dev
-------------
 - Bugs Fix
 - Public Area (File Upload) Works
 - Working Submit Validation Works

v20150411 dev
-------------
 - Bugs Fix
 - Working Preview Function Works

v20150410 dev
-------------
 - Bugs Fix
 - New Management Office Reaction Header
 - New Management Lobby Office Page UI

v20150409 dev
-------------
 - Bugs Fix
 - New Reply Style
 - Reply GitHub Flavored Markdown Support
 - New Checkin Page UI

v20150408 dev
-------------
 - Bugs Fix
 - Gravatar Cache Function Works
 - Reply Function Works

v20150407 dev
-------------
 - Bugs Fix
 - New UI

v20150406 dev
-------------
 - Bugs Fix
 - Misaka GitHub Flavored Markdown works

v20150404 dev
-------------
 - Bugs Fix
 - Removed jQuery file and Starts to Use Microsoft Ajax Content Delivery Network
 - Index Page Framework Finished

v20150403 dev
-------------
 - Bugs Fix
 - Publish Function Works
 - CRDA Page Added

v20150402 dev
-------------
 - Slug Feature Added
 - Working Page JavaScript Post Function Completed

v20150401 dev
-------------
 - Management Enhance
 - Changed Template System to Mako
 - Completed Apache License Comments
 - Bugs Fix
 - Font Family Update
 - Footer Added

v20150331 dev
-------------
 - Bugs Fix
 - Checkout function works

v20150330 dev
-------------
 - Project initialized.


foundation/office.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
#   Copyright 2015 Futur Solo
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.


from tornado.web import *
from tornado.gen import *
from foundation.place import PlacesOfInterest, slug_validation, visitor_only
from collections import OrderedDict
import pyotp
import json
import bcrypt
import datetime
import time


class ManagementOffice(PlacesOfInterest):
    def management_url(self, path, include_host=None, **kwargs):
        path = "management/" + path
        return RequestHandler.static_url(
            self, path, include_host=include_host, **kwargs)

    def management_render(self, page):
        self.render_list["management_url"] = self.management_url
        page = "management/" + page
        self.render(page, nutrition=False)


class CheckinOffice(ManagementOffice):
    @coroutine
    @visitor_only
    def get(self):
        self.render_list["checkin_status"] = self.get_scookie(
            "checkin_status", arg_type="hash", default="ok")
        self.clear_cookie("checkin_status")
        self.management_render("checkin.htm")

    @coroutine
    @visitor_only
    def post(self):
        username = self.get_arg("username", arg_type="username")
        password = self.get_arg("password", arg_type="hash")
        two = self.get_arg("two", arg_type="number")
        remember = self.get_arg("remember", arg_type="boolean")

        if not (username and password and two):
            self.set_scookie("checkin_status", "password", expires_days=None)
            self.redirect("/management/checkin")
            return

        user = yield self.get_user(username=username)

        """
        generate new password by listed commands:
        password = "YOUR NEW PASSWORD"
        bcrypt.hashpw(
            hashlib.sha256(password.encode()
                           ).hexdigest().encode(), bcrypt.gensalt()
        )
        """

        if not user:
            self.set_scookie("checkin_status", "password", expires_days=None)
            self.redirect("/management/checkin")
            return

        if bcrypt.hashpw(
           password.encode("utf-8"),
           user["password"].encode()) != user["password"].encode():
            self.set_scookie("checkin_status", "password", expires_days=None)
            self.redirect("/management/checkin")
            return

        def verify_otp(key, two):
            totp = pyotp.TOTP(key)
            current_datetime = datetime.datetime.now()
            if totp.verify(two, for_time=current_datetime):
                return True
            early_datetime = current_datetime - datetime.timedelta(seconds=30)
            if totp.verify(two, for_time=early_datetime):
                return True
            later_datetime = current_datetime + datetime.timedelta(seconds=30)
            if totp.verify(two, for_time=later_datetime):
                return True
            return False

        if not verify_otp(user["otp_key"], two):
            self.set_scookie("checkin_status", "two", expires_days=None)
            self.redirect("/management/checkin")
            return

        else:
            expires_days = None
            if remember:
                expires_days = 180

            device_id = self.get_random(32)

            agent_auth = self.hash((device_id + user["password"]), "sha256")

            self.set_scookie("user_id", user["_id"], expires_days=expires_days,
                             httponly=True)

            self.set_scookie("device_id", device_id, expires_days=expires_days,
                             httponly=True)

            self.set_scookie("agent_auth", agent_auth,
                             expires_days=expires_days, httponly=True)

            self.redirect(self.next_url)


class CheckoutOffice(ManagementOffice):
    @authenticated
    def get(self):
        self.clear_cookie("user_id")
        self.clear_cookie("device_id")
        self.clear_cookie("agent_auth")
        self.redirect(self.next_url)


class MainOffice(ManagementOffice):
    @coroutine
    @authenticated
    def get(self, slug, sub_slug=""):
        if not self.value_validation("hash", slug):
            raise HTTPError(404)
        if sub_slug and not self.value_validation("hash", sub_slug):
            raise HTTPError(404)
        self.render_list["slug"] = slug
        self.render_list["sub_slug"] = sub_slug
        self.management_render("office.htm")


class ActionOffice(ManagementOffice):
    @coroutine
    @authenticated
    def post(self):
        action = self.get_arg("action", default=None, arg_type="link")
        if hasattr(self, action):
            yield getattr(self, action)()
        else:
            raise HTTPError(500)

    @coroutine
    def load_public(self):
        book = self.memories.select("Publics")
        book.find({"type": "file"}).sort([["time", False]])
        book.length(0, force_dict=True)
        yield book.do()
        result = book.result()
        self.finish(json.dumps(list(result.values())))

    @coroutine
    def save_public(self):
        public_path = os.path.join(
            os.path.join(
                self.settings["static_path"], "public"), "files")
        url_base = "/spirit/public/files"

        if self.request.files:
            for f in self.request.files["files[]"]:
                book = self.memories.select("Publics")
                current_time = int(time.time())
                current_path = os.path.join(public_path, str(
                    current_time))
                current_url = os.path.join(url_base, str(
                    current_time))
                if not os.path.exists(current_path):
                    os.makedirs(current_path)

                filename = f["filename"]
                current_file_path = os.path.join(
                    current_path, filename)
                current_file_url = os.path.join(
                    current_url, filename)

                with open(current_file_path, "wb") as file:
                    file.write(f["body"])

                file_info = OrderedDict()
                file_info["time"] = current_time
                file_info["type"] = "file"
                file_info["content_type"] = None
                file_info["filename"] = filename
                file_info["filepath"] = current_file_path
                file_info["fileurl"] = current_file_url
                file_info["email_md5"] = None
                file_info["_id"] = yield self.issue_id("Publics")
                book.add(file_info)
                yield book.do()
        else:
            raise HTTPError(500)
        self.finish(json.dumps({"status": True}))

    @coroutine
    def count(self):
        info = yield self.get_count()
        self.finish(json.dumps(info))

    @coroutine
    def save_working(self):
        working_type = self.get_arg("working_type", arg_type="hash")
        if working_type == "writing":
            book = self.memories.select("Writings")
        elif working_type == "page":
            book = self.memories.select("Pages")
        else:
            raise HTTPError(500)

        working_method = self.get_arg("working_method", arg_type="hash")

        working_id = self.get_arg("working_id", arg_type="number")

        def make_working():
            working = {}
            working["title"] = self.get_arg("working_title", arg_type="origin")
            working["content"] = self.get_arg("working_content",
                                              arg_type="origin")
            working["time"] = self.get_arg("working_time", arg_type="number")
            working["publish"] = self.get_arg("working_publish",
                                              arg_type="boolean")
            working["slug"] = self.get_arg("working_slug", arg_type="slug")
            working["author"] = self.current_user["_id"]
            if not working["slug"]:
                raise HTTPError(500)
            return working

        def check_slug(slug):
            book.find({"slug": slug})
            yield book.do()
            slug_result = book.result()
            if slug_result and (
             slug_result is not False and slug_result["_id"] != working_id):
                self.finish(json.dumps({"succeed": False, "reason": "slug"}))
                return False
            return True

        if working_method == "new":
            working = make_working()
            if not check_slug(working["slug"]):
                return
            if working_type == "writing":
                working_id = yield self.issue_id("Writings")
            elif working_type == "page":
                working_id = yield self.issue_id("Pages")
            else:
                raise HTTPError(500)
            working["_id"] = working_id
            book.add(working)
        elif working_method == "edit":
            working = make_working()
            if not check_slug(working["slug"]):
                return
            book.set({"_id": working_id}, working)
        elif working_method == "erase":
            @coroutine
            def erase_reply(working_id):
                book = self.memories.select("Replies")
                book.erase({"writing_id": working_id})
                yield book.do()
            if working_type == "writing":
                yield erase_reply(working_id)
            book.erase({"_id": working_id})
        else:
            raise HTTPError(500)
        yield book.do()
        self.finish(json.dumps({
            "succeed": True,
            "id": working_id,
        }))

    @coroutine
    def load_working(self):
        working_type = self.get_arg("type", arg_type="hash")
        working_id = self.get_arg("id", arg_type="number")
        if working_type == "writing":
            book = self.memories.select("Writings")
        elif working_type == "page":
            book = self.memories.select("Pages")
        else:
            raise HTTPError(500)
        book.find({"_id": working_id})
        yield book.do()
        working = book.result()
        self.finish(json.dumps(working))

    @coroutine
    def load_crda(self):
        type = self.get_arg("type", arg_type="hash")

        if type == "writings":
            book = self.memories.select("Writings")
            book.find({}, ["content"])
        elif type == "pages":
            book = self.memories.select("Pages")
            book.find({}, ["content"])
        elif type == "replies":
            book = self.memories.select("Replies")
            book.find({})
            writing_list = []
        else:
            raise HTTPError(500)

        book.sort([["time", False]])
        book.length(0, True)
        yield book.do()
        content_list = book.result()

        if type == "replies":
            for key in content_list:
                content_list[key]["_id"] = int(
                    content_list[key]["_id"])
                if content_list[key]["writing_id"] not in writing_list:
                    writing_list.append(content_list[key]["writing_id"])

            writing_list = yield self.get_writing(writing_list=writing_list)
            for key in content_list:
                if content_list[key]["writing_id"] not in writing_list.keys():
                    del content_list[key]
                    continue
                content_list[key]["writing"] = writing_list[
                    content_list[key]["writing_id"]]

        self.finish(json.dumps(list(content_list.values())))

    @coroutine
    def save_reply(self):
        reply_id = self.get_arg("reply", arg_type="number")
        reply_method = self.get_arg("method", arg_type="hash")
        book = self.memories.select("Replies")

        if reply_method == "permit":
            permit = self.get_arg("permit", arg_type="boolean")
            if permit is None:
                raise HTTPError(500)

            book.find({"_id": reply_id})
            yield book.do()
            reply = book.result()
            if not reply:
                raise HTTPError(500)

            book.set({"_id": reply_id}, {"permit": permit})
            yield book.do()
            self.finish(json.dumps({"status": True}))

        elif reply_method == "erase":
            book.erase({"_id": reply_id})
            yield book.do()
            self.finish(json.dumps({"status": True}))

        elif reply_method == "edit":
            reply_name = self.get_arg("name", arg_type="origin")
            reply_homepage = self.get_arg("homepage", arg_type="origin")
            reply_email = self.get_arg("email", arg_type="mail_address")
            reply_content = self.get_arg("content", arg_type="origin")
            if not (reply_id and reply_name and reply_homepage and
                    reply_email and reply_content):
                raise HTTPError(500)
            reply = {}
            reply["name"] = reply_name
            reply["homepage"] = reply_homepage
            reply["email"] = reply_email
            reply["content"] = reply_content
            book.set({"_id": reply_id}, reply)
            yield book.do()
            self.finish(json.dumps({"status": True}))

    @coroutine
    def load_configuration(self):
        book = self.memories.select("Configs")
        book.find({})
        book.length(0, True)
        yield book.do()
        configs = book.result()
        self.finish(json.dumps(configs))

    @coroutine
    def save_configuration(self):
        post_config = OrderedDict()
        post_config["site_name"] = self.get_arg("site_name", arg_type="origin")
        post_config["site_description"] = self.get_arg(
            "site_description", arg_type="origin")
        post_config["site_keywords"] = self.get_arg(
            "site_keywords", arg_type="origin")
        post_config["site_url"] = self.get_arg("site_url", arg_type="link")
        post_config["nutrition_type"] = self.get_arg(
            "nutrition_type", arg_type="hash")
        post_config["trace_code"] = self.get_arg(
            "trace_code", arg_type="origin")
        for key in post_config:
            if not post_config[key]:
                raise HTTPError(500)
        book = self.memories.select("Configs")
        book.find({}).length(0, force_dict=True)
        yield book.do()
        origin_config = book.result()
        for key in post_config:
            if origin_config[key] != post_config[key]:
                book.set({"_id": key}, {"value": post_config[key]})
                yield book.do()
        self.finish(json.dumps({"status": True}))
